### PR TITLE
Feat pipeline linear regression

### PR DESCRIPTION
**Pull request by papapana**
Code review for youtube_analysis_workflow.py

### CODE REVIEW

File: youtube_comedy_slam/src/data/make_dataset.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+
+def clean_google_dataset(full_dataset, google_dataset_old, google_dataset_new):
+    """
+    Cleans the google_dataset_old of youtube ids that do not exist in the full_dataset
+    and writes the result in google_dataset_new
+    :param full_dataset: the full downloaded dataset
+    :param google_dataset_old: the filename of the google dataset (format id1,id2,(left|right)
+    :param google_dataset_new: the filename of the clean dataset
+    :return:
```
Comment (ziky90) [inline youtube_comedy_slam/src/data/make_dataset.py:30]: When default None is returned just delete this line.

File: youtube_comedy_slam/src/data/make_dataset.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    """
+
+    # Reading a big csv results in reading limitations of the field limit in the csv module
```
Comment (ziky90) [inline youtube_comedy_slam/src/data/make_dataset.py:33]: Think of storing data in several smaller files or some more suitable format.

File: youtube_comedy_slam/src/data/make_dataset.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+        decrement = False
+        try:
+            csv.field_size_limit(maxInt)
+        except OverflowError:
+            maxInt = int(maxInt / 10)
+            decrement = True
+
+    csv.field_size_limit(maxInt)
+
+    youtube_ids = {}
```
Comment (ziky90) [inline youtube_comedy_slam/src/data/make_dataset.py:52]: You can use set instead of the dict.

File: youtube_comedy_slam/src/data/make_dataset.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+            records_before = num_of_lines(google_dataset_old)
+            print("Cleaning dataset of inexisting names...")
+            clean_google_dataset(full_dataset, google_dataset_old, google_dataset_new)
+            records_new = num_of_lines(google_dataset_new)
+            print("Dataset records before:", records_before, "Dataset records now:", records_new)
+        except Exception as err:
+            logging.error("Unexpected error:", sys.exc_info()[0], "Exception", err)
+
+
+
```
Comment (ziky90) [inline youtube_comedy_slam/src/data/make_dataset.py:100]: too many empty lines at the end of the file

File: youtube_comedy_slam/src/data/make_dataset.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+            new_file.write("video_id1,video_id2,funnier")
+
+            for row in old_ds:
+                if row[0] in youtube_ids and row[1] in youtube_ids:
+                    new_file.write(",".join(row) + "\n")
+
+
+if __name__ == '__main__':
+    log_fmt = '%(asctime)s - %(name)s - %(levelname)s - %(message)s'
+    logging.basicConfig(level=logging.INFO, format=log_fmt)
```
Comment (ziky90) [inline youtube_comedy_slam/src/data/make_dataset.py:70]: maybe logging also to the file might be helpful for later debugging.

File: youtube_comedy_slam/src/data/make_dataset.py [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+    load_dotenv(find_dotenv())
+    PARSER = argparse.ArgumentParser()
+    SUB_PARSER = PARSER.add_subparsers(help='Command to execute', title='actions', dest='action')
+    SUB_PARSER.required = True
+    SP = SUB_PARSER.add_parser('clean_google_dataset', help='Will clean the published google dataset')
+    SP.add_argument('--old-file', type=argparse.FileType('r'), help='Old google dataset', required=True)
+    SP.add_argument('--new-file', type=argparse.FileType('w'), help='New google dataset', required=True)
+    SP.add_argument('--full-dataset', type=argparse.FileType('r'), help='Full dataset', required=True)
+    ARGS = PARSER.parse_args()
+    if ARGS.action == 'clean_google_dataset':
```
Comment (ziky90) [inline youtube_comedy_slam/src/data/make_dataset.py:86]: this looks like only option so far, so maybe parser can be simplified + this condition can be omitted.

File: youtube_comedy_slam/src/data/merge_new.sh [hunk 7]
```diff
@@ -0,0 +1,6 @@
+#!/usr/bin/env sh
+
+NEWFILE="dataset_movies.${1}_${2}_full.csv"
+mv dataset_movies.csv $NEWFILE
+PREV=`expr $1 - 1`
+cat "dataset_movies.1_${PREV}_full.csv" "${NEWFILE}" > "dataset_movies.1_${2}_full.csv"
```
Comment (ziky90) [inline youtube_comedy_slam/src/data/merge_new.sh:6]: newline at the end of the file

File: youtube_comedy_slam/src/data/merge_new_test.sh [hunk 8]
```diff
@@ -0,0 +1,6 @@
+#!/usr/bin/env sh
+
+NEWFILE="test_dataset_movies.${1}_${2}_full.csv"
+mv test_dataset_movies.csv $NEWFILE
+PREV=`expr $1 - 1`
+cat "test_dataset_movies.1_${PREV}_full.csv" "${NEWFILE}" > "test_dataset_movies.1_${2}_full.csv"
```
Comment (ziky90) [inline youtube_comedy_slam/src/data/merge_new_test.sh:6]: newline at the end of the file

File: youtube_comedy_slam/src/data/youtube_comments_api/youtube_comments_api.py [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+import sys
+
+from apiclient.discovery import build
+from apiclient.errors import HttpError
+from tqdm import tqdm
+
+
+def get_comment_threads(youtube_service, video_id, csvfile, max_results=100, limit_pages=None, write_header=True):
+    """
+    Our purpose is to download comments and append them to a csv with the following format:
```
Comment (ziky90) [inline youtube_comedy_slam/src/data/youtube_comments_api/youtube_comments_api.py:21]: Docstrings are usualy in a following form:
```
Download comments and append them to a csv with the following format:
.....
```

File: youtube_comedy_slam/src/data/youtube_comments_api/youtube_comments_api.py [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+    videoId,textDisplay,isReplyTo,canReply,totalReplyCount,etag,id,authorChannelId,authorDisplayName,likeCount,
+    publishedAt, updatedAt,viewerRating
+    :param youtube_service: youtube object from the Youtube API, see get_default_youtube_service
+    :param video_id: youtube video_id to download
+    :param csvfile: the resulting csvfile to write
+    :param max_results: max_results for each call (1-100)
+    :param limit_pages: if set limits the comment pages to download, mainly for testing purposes
+    :param write_header: write or not the csv header
+    :return: top results but also write everything in the csvfile
```
Comment (ziky90) [inline youtube_comedy_slam/src/data/youtube_comments_api/youtube_comments_api.py:30]: Information about writing to csv shouldn't be here, but in the general docstring.

File: youtube_comedy_slam/src/data/youtube_comments_api/youtube_comments_api.py [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+
+    # Setup csv file
+    fieldnames = ['videoId', 'textDisplay', 'isReplyTo', 'canReply', 'totalReplyCount', 'kind', 'etag', 'id',
+                  'authorChannelId', 'authorDisplayName', 'likeCount', 'publishedAt', 'updatedAt', 'viewerRating']
+    with open(csvfile, 'a') as fcsv:
+        writer = csv.DictWriter(fcsv, fieldnames=fieldnames)
+        if write_header:
+            writer.writeheader()
+
+        def get_comment_page(page_results, ignore_first=False):
```
Comment (ziky90) [inline youtube_comedy_slam/src/data/youtube_comments_api/youtube_comments_api.py:50]: nested methods are possible to use, but I recommend to put them always on the top of your function.

File: youtube_comedy_slam/src/data/youtube_comments_api/youtube_comments_api.py [hunk 12]
```diff
@@ -0,0 +1,197 @@
...
+            """
+            Processes a comment_page of results
+            :param page_results: The results of a comment page
+            :param ignore_first: Ignore the first result to avoid duplicates from previous calls
+            :return:
```
Comment (ziky90) [inline youtube_comedy_slam/src/data/youtube_comments_api/youtube_comments_api.py:55]: this line can be deleted

File: youtube_comedy_slam/src/data/youtube_comments_api/youtube_comments_api.py [hunk 13]
```diff
@@ -0,0 +1,197 @@
...
+import os
+import socket
+import sys
+
+from apiclient.discovery import build
+from apiclient.errors import HttpError
+from tqdm import tqdm
+
+
+def get_comment_threads(youtube_service, video_id, csvfile, max_results=100, limit_pages=None, write_header=True):
```
Comment (ziky90) [inline youtube_comedy_slam/src/data/youtube_comments_api/youtube_comments_api.py:19]: generally speaking this looks too complicated, thinking of some 3rd party library for parsing or at least splitting this to somme smaller methods/object might be beneficial for later modifications or debugging.

File: youtube_comedy_slam/src/features/build_features.py [hunk 14]
```diff
@@ -0,0 +1,47 @@
+#!/usr/bin/env python3
+
+# -*- coding: utf-8 -*-
+import argparse
+
+def feature_sum_score(dataset_df, normalized=True):
+    """
+    This functions creates a score for each video which is the sum(better)-sum(worse)
+    :param dataset_df: The dataset panda dataframe
```
Comment (ziky90) [inline youtube_comedy_slam/src/features/build_features.py:9]: Missing parameter `normalized` description. At least  providing the info that it'll be divided my maximum.

File: youtube_comedy_slam/src/features/build_features.py [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+    better2 = dataset_df[dataset_df.funnier == 'right'].video_id2.sort_index().value_counts()
+    better = better1.add(better2, fill_value=0).sort_index()
+    worse1 = dataset_df[dataset_df.funnier == 'right'].video_id1.sort_index().value_counts()
+    worse2 = dataset_df[dataset_df.funnier == 'left'].video_id2.sort_index().value_counts()
+    worse = worse1.add(worse2, fill_value=0).sort_index()
+    score = better.subtract(worse, fill_value=0).sort_index()
+    if normalized:
+        max_score = score.abs().max()
+        assert max_score != 0
+        score = score / max_score
```
Comment (ziky90) [inline youtube_comedy_slam/src/features/build_features.py:22]: Storing the maximum would be helpful for the later denormalization.

File: youtube_comedy_slam/src/tasks/LinearRegressionScorePerVideoTask.py [hunk 16]
```diff
@@ -0,0 +1,23 @@
...
+        # Train the model using the training sets
+        regr.fit(scipy_train.transpose(), label)
+
+    def output(self):
+        return luigi.LocalTarget('/tmp/foo/bar-%s.txt' % self.param)
+
+
+
+if __name__ == '__main__':
+    luigi.run()
```
Comment (ziky90) [inline youtube_comedy_slam/src/tasks/LinearRegressionScorePerVideoTask.py:23]: missing newline at the end of the file

File: youtube_comedy_slam/src/tasks/youtube_analysis_workflow.py [hunk 17]
```diff
@@ -0,0 +1,352 @@
...
+        exists to determine whether the Task needs to run at all.
+        """
+        return {'train': luigi.LocalTarget(root_path + 'data/processed/train_dataset_with_lang_feature.csv'),
+                'test': luigi.LocalTarget(root_path + 'data/processed/test_dataset_with_lang_feature.csv')}
+
+    def run(self):
+        """
+        AddLangFeatureTask :
+        """
+
```
Comment (ziky90) [inline youtube_comedy_slam/src/tasks/youtube_analysis_workflow.py]: This wouldn't compile. So at least temporarily `pass` would be good.

File: youtube_comedy_slam/src/test/test_get_comment_threads.py [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+
+        # Now check if the replies are parsed
+        results = get_comment_threads(youtube_service=youtube, video_id="-WVvOU1r2sM", csvfile="data/test.csv",
+                                      max_results=10, limit_pages=1)
+        mystr = open('data/test.csv').read()
+        assert 'z12xjrcijsr4jnqhp04chxcptlbtjrn5zz00k.1473936165253594' in mystr
+
+
+if __name__ == '__main__':
+    unittest.main()
```
Comment (ziky90) [inline youtube_comedy_slam/src/test/test_get_comment_threads.py:43]: newline at the end of the file

File: youtube_comedy_slam/tox.ini [hunk 19]
```diff
@@ -0,0 +1,3 @@
+[flake8]
+max-line-length = 79
+max-complexity = 10
```
Comment (ziky90) [inline youtube_comedy_slam/tox.ini:3]: newline at the end of the file

**Review by ziky90 — CHANGES_REQUESTED** [review]
Here are few general comments:

1. Try to make smaller pull requests more frequently (it's much easier for reviewer and for fixing bugs)

2. Whole repo looks kind of messy, some general `README.md` file with usage description would be very helpful.
